### PR TITLE
Stc 10

### PR DESCRIPTION
**Pull request by JKFerland**
See latest commit notes for updated functionality. I pulled in STC-11 to match indentation, but let me know if any files from STC-10 missed my review :-).

### CODE REVIEW

File: .project [hunk 1]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (jdamerow) [inline .project:1]: let's not add this file

File: scriptoocloud/pom.xml [hunk 2]
```diff
@@ -19,14 +19,18 @@
 		<admin.password>$2a$04$oQo44vqcDIFRoYKiAXoNheurzkwX9dcNmowvTX/hsWuBMwijqn44i</admin.password>
 
 		<db.driver>com.mysql.jdbc.Driver</db.driver>
-		<db.database.url>jdbc-url-to-db</db.database.url>
+		<db.database.url>jdbc:mysql://localhost:3306/STCDB?useSSL=false&amp;allowPublicKeyRetrieval=true</db.database.url>
```
Comment (jdamerow) [inline scriptoocloud/pom.xml:22]: let's not change this

File: scriptoocloud/pom.xml [hunk 3]
```diff
@@ -238,11 +242,32 @@
...
 			<artifactId>org.eclipse.jgit</artifactId>
 			<version>3.5.0.201409260305-r</version>
 		</dependency>
+		
+		<!-- https://mvnrepository.com/artifact/org.yaml/snakeyaml -->
+		<dependency>
+		    <groupId>org.yaml</groupId>
+		    <artifactId>snakeyaml</artifactId>
+		    <version>1.28</version>
+		</dependency>
```
Comment (jdamerow) [inline scriptoocloud/pom.xml:270]: the above changes are not from your story, are they?

File: scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/ProjectManager.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+public interface ProjectManager {
+
+    Project createProject(String name, String description, IUser user);
+
+    void deleteProject(Long id);
+
+    List<Project> findAll();
+
+	Optional<ProjectImpl> getProject(int id);
```
Comment (jdamerow) [inline scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/ProjectManager.java:18]: indentation

File: scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/ProjectManagerImpl.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+	public List<Project> findAll() {
+		Iterable<ProjectImpl> projects = projectRepo.findAll();
+		List<Project> results = new ArrayList<>();
+		projects.forEach(e -> results.add((Project) e));
+		return results;
+	}
+
+	@Override
+	public void updateProject(int id, String name, String description) {
+		//project.setId(id);
```
Comment (jdamerow) [inline scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/ProjectManagerImpl.java:78]: remove

File: scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/ProjectManagerImpl.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+	    Optional<ProjectImpl> project = getProject(id);
+	    Project project1 = project.get();
```
Comment (jdamerow) [inline scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/ProjectManagerImpl.java:80]: you need to check first if the project exists

File: scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/ProjectManagerImpl.java [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+		project1.setName(name);
+		project1.setDescription(description);
+		project1.setLastModifiedDate(ZonedDateTime.now());
+
+		projectRepo.save((ProjectImpl) project1);
```
Comment (jdamerow) [inline scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/ProjectManagerImpl.java:85]: indentation

File: scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/ProjectManagerImpl.java [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+	}
+
+
+    @Override
+    public Optional<ProjectImpl> getProject(int id) {
+        Optional<ProjectImpl> project = Optional.ofNullable(new ProjectImpl());
```
Comment (jdamerow) [inline scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/ProjectManagerImpl.java:91]: what does this line do?

File: scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/ProjectManagerImpl.java [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+		project1.setName(name);
+		project1.setDescription(description);
+		project1.setLastModifiedDate(ZonedDateTime.now());
+
+		projectRepo.save((ProjectImpl) project1);
+	}
+
+
+    @Override
+    public Optional<ProjectImpl> getProject(int id) {
```
Comment (jdamerow) [inline scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/ProjectManagerImpl.java:90]: your should return IProject here.

File: scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/AddProjectController.java [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+public class AddProjectController {
+
+    @Autowired
+    private ProjectManager projectManager;
+
+    @Autowired
+    private final IUserManager userManager;
+
+    @Autowired
+    public AddProjectController(IUserManager userManager) {
```
Comment (jdamerow) [inline scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/AddProjectController.java:34]: either annotate constructor or class variables but not both (consistency)

File: scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/AddProjectController.java [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+    public String get(Model model) {
+        model.addAttribute("project", new ProjectImpl());
+        return "/auth/add";
+    }
+
+    @RequestMapping(value = "/auth/add", method = RequestMethod.POST)
+    public String post(@Valid @ModelAttribute("project") ProjectImpl projectImpl, BindingResult result, Model model,
+            RedirectAttributes redirectAttrs, Principal principal) {
+        if (result.hasErrors()) {
+            model.addAttribute("project", projectImpl);
```
Comment (jdamerow) [inline scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/AddProjectController.java:48]: this won't do much if you redirect later.

File: scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/EditProjectController.java [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+import edu.asu.diging.scriptoocloud.core.model.impl.ProjectImpl;
+import edu.asu.diging.scriptoocloud.core.service.ProjectManager;
+
+@Controller
+public class EditProjectController {
+
+    @Autowired
+    private ProjectManager projectManager;
+    
+    private int projectID;
```
Comment (jdamerow) [inline scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/EditProjectController.java:32]: your controller is in singleton scope so all request share this variable

File: scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/EditProjectController.java [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+
+    @RequestMapping(value = "/auth/edit/{id}", method = RequestMethod.POST)
+    public String post(@Valid @ModelAttribute("project") Optional<ProjectImpl> projectImpl, BindingResult result,
+            Model model, RedirectAttributes redirectAttrs, Principal principal, @PathVariable("id") int id) {
+        if (result.hasErrors()) {
+            model.addAttribute("project", new ProjectImpl());
+            return "/auth/edit/{id}";
+        }
+        projectImpl = projectManager.getProject(id);
+        projectID = id;
```
Comment (jdamerow) [inline scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/EditProjectController.java:42]: remove

File: scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/EditProjectController.java [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+        model.addAttribute("project", projectImpl);
+        return "/auth/edit";
+    }
+
+    @RequestMapping(value = "/auth/edit", method = RequestMethod.POST)
```
Comment (jdamerow) [inline scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/EditProjectController.java:47]: add project id to url

File: scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/EditProjectController.java [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+@Controller
+public class EditProjectController {
+
+    @Autowired
+    private ProjectManager projectManager;
+    
+    private int projectID;
+
+    @RequestMapping(value = "/auth/edit/{id}", method = RequestMethod.POST)
+    public String post(@Valid @ModelAttribute("project") Optional<ProjectImpl> projectImpl, BindingResult result,
```
Comment (jdamerow) [inline scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/EditProjectController.java:35]: should be a get request

File: scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/EditProjectController.java [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+        projectID = id;
+        model.addAttribute("project", projectImpl);
+        return "/auth/edit";
+    }
+
+    @RequestMapping(value = "/auth/edit", method = RequestMethod.POST)
+    public String post2(@Valid @ModelAttribute("project") ProjectImpl projectImpl, BindingResult result,
+            Model model, RedirectAttributes redirectAttrs, Principal principal) {
+        ProjectImpl project = projectImpl;
+        projectManager.updateProject((int) projectID, project.getName(), project.getDescription());
```
Comment (jdamerow) [inline scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/EditProjectController.java:51]: get project id from path

File: scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/ProjectsController.java [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+    private ProjectRepository projectRepo;
+
+    @Autowired
+    private RemoveProjectController removeProjectController;
+
+    @Autowired
+    private AddProjectController addProjectController;
+
+    @Autowired
+    private ProjectManager projectManager;
```
Comment (jdamerow) [inline scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/ProjectsController.java:34]: remove unnecessary dependencies

**Review by jdamerow — CHANGES_REQUESTED** [review]
Your indentation settings are still wrong. Also, please reopen this PR next time to keep the comments.

File: scriptoocloud/pom.xml (hunk 3 above)
Comment (JKFerland) [reply scriptoocloud/pom.xml:270]: I merged in develop and I believe this is needed for some of Jasons dependencies from stories 5/15

File: scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/EditProjectController.java (hunk 15 above)
Comment (JKFerland) [reply scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/EditProjectController.java:35]: Do you have an example I may look at that utilizes a GET and POST? Currently, it POST's from the button press, then again when the user submits the information. I had some trouble understanding how I could do the same/similar thing ( have the user access edit.html with the projectID as a path variable) with a GET/POST exchange.

File: scriptoocloud/pom.xml (hunk 2 above)
Comment (JKFerland) [reply scriptoocloud/pom.xml:22]: Fix reflected in STC-11 and merged into STC-10

File: .project (hunk 1 above)
Comment (JKFerland) [reply .project:1]: Will delete in next commit via Github.

File: scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/ProjectManagerImpl.java (hunk 6 above)
Comment (JKFerland) [reply scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/ProjectManagerImpl.java:80]: Ok. Should I just display a message to the user that the project does not exist? This is similar to the delete project button, and I dont think we should throw an exception like we didn't in the RemoveProjectController(if that makes sense).

File: scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/ProjectManagerImpl.java (hunk 7 above)
Comment (JKFerland) [reply scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/ProjectManagerImpl.java:85]: Fixed with new indentation rules.

File: scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/ProjectManagerImpl.java (hunk 8 above)
Comment (JKFerland) [reply scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/ProjectManagerImpl.java:91]: The project repository only returns an optional ProjectImpl when using findById (Line 92), so I think I can only update projects using an Optional ProjectImpl. Is this assumption wrong? Can I convert from the optional to a normal version of the object?

File: scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/ProjectManagerImpl.java (hunk 9 above)
Comment (JKFerland) [reply scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/ProjectManagerImpl.java:90]: What is an IProject? Should I not be using my ProjectImpl implementation?

File: scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/AddProjectController.java (hunk 10 above)
Comment (JKFerland) [reply scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/AddProjectController.java:34]: IUserManager annotated in STC-11, merged into STC-10 so change is now reflected.

File: scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/ProjectsController.java (hunk 17 above)
Comment (JKFerland) [reply scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/ProjectsController.java:34]: Removed in STC-11 and change reflected in STC-10

File: scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/ProjectManagerImpl.java (hunk 6 above)
Comment (jdamerow) [reply scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/ProjectManagerImpl.java:80]: Yes, that works.

File: scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/ProjectManagerImpl.java (hunk 9 above)
Comment (jdamerow) [reply scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/ProjectManagerImpl.java:90]: Or `Project` (whatever the interface that `ProjectImpl` implements is called).

File: scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/ProjectManagerImpl.java (hunk 8 above)
Comment (jdamerow) [reply scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/ProjectManagerImpl.java:91]: you are replacing `project` with the result of `findById`, so this line doesn't do anything. Optionals are just a more convenient way to deal with null values. You can use their `isPresent` and `get` methods to get the object you are actually interested in.

File: scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/EditProjectController.java (hunk 14 above)
Comment (JKFerland) [reply scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/EditProjectController.java:47]: Added

**Pull request closed without merging**